dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFileHandler.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.opensource.classpath;
+
+import com.google.common.collect.ImmutableList;
+import java.util.ArrayDeque;
+import org.xml.sax.Attributes;
+import org.xml.sax.SAXException;
+import org.xml.sax.helpers.DefaultHandler;
+
+/**
+ * Handler for Linkage Checker exclusion XML files.
+ *
+ * <p>An instance of this class can process one document and cannot be reused for other documents.
+ */
+class ExclusionFileHandler extends DefaultHandler {
+
+  /**
+   * Stack to hold the ancestors of the XML element in {@link #startElement(String, String, String,
+   * Attributes)}.
+   */
+  private ArrayDeque<SymbolProblemMatcher> stack = new ArrayDeque<>();
+
+  private ImmutableList.Builder<LinkageErrorMatcher> matchers;
+
+  ImmutableList<LinkageErrorMatcher> getMatchers() {
+    return matchers.build();
+  }
+
+  @Override
+  public void startDocument() {
+    if (matchers != null) {
+      throw new IllegalStateException("This handler started reading document already");
+    }
+    matchers = ImmutableList.builder();
+  }
+
+  /** Adds {@code child} to the parent matcher that is at the top of the {@link #stack}. */
+  private void addMatcherToParent(SymbolProblemTargetMatcher child) throws SAXException {
+    SymbolProblemMatcher parent = stack.peek();
+    if (parent instanceof SourceMatcher && child instanceof SymbolProblemSourceMatcher) {
```

Comment:
This is an awful lot of casting and instanceof, and it makes me wonder if we're not taking advantage of polymorphism where we could, or perhaps we don't have quite the right class hierarchy.

Is there perhaps some way SymbolProblemMatcher.setMatcher could simply work when it's passed any instance of SymbolProblemTargetMatcher?

Reply:
> SymbolProblemMatcher.setMatcher could simply work when it's passed any instance of SymbolProblemTargetMatcher?

No, SourceMatcher cannot take FieldMatcher. I replaced the stack and this function with the two variables linkageErrorMatcher and symbolProblemMatcher. Now the logic is much simpler. Does this look good now?

Reply:
SymbolProblemMatcher's addChild now may be the one you mentioned above:

```
  default void addChild(SymbolProblemTargetMatcher child) {
    throw new IllegalStateException("This element is not supposed to take child element");
  }
```